### ISSUE TITLE
Can't see the multiline layout

### ISSUE THREAD

**Issue by ynjo** [Betterbird/thunderbird-patches#9]
Hello,

I installed the multine preview, but in the vertical layout I still see one line per message. Is there a way to activate the multiline view?

Thanks!

**Comment by Betterbird** [comment]
First of all, use "preview 4", the previous ones had their quirks (and this one still has a quirk as I just found). Then, in the thread pane/message list go to the column picker and switch off the subject column:
![image](https://user-images.githubusercontent.com/86832723/132772542-62a3c53c-4171-4192-9f14-9faa4b0f8ab2.png)
We might be doing this differently in the future.

**Comment by ynjo** [comment]
Thanks for the immediate answer, and thank you very much for addressing such a long awaited feature!
Some thoughts:

- about the per-folder option, I guess it's a matter of taste, but for example in my case, working on many projects and having many folders, it would help a lot to have a global option somewhere to avoid making the same changes in every folder. Maybe it could be a layout option, since I guess that users who like this view would use it everywhere.

- in a multiline view, I think we need a better way to discern the different emails among lines. I see that the second line is italic and that's good, maybe the sender could be bold? I saw that in other email clients with multiline view. I see you use alternate colors for every email line, which helps too. Is it thanks to user CSS file or to some extension you installed?

**Comment by Betterbird** [comment]
Point 1: per-folder vs. global option. Well, doing the per-folder thing was quite hard and it's still not perfect. I global option would have been much easier. Personally, I can't see why I would want to use the feature in certain folders. I guess a global option will be coming, and it won't take 18 years (https://bugzilla.mozilla.org/show_bug.cgi?id=213945). But I need to see how to integrate it with the per-folder option. You know that you can propagate folder settings to a folder and all its subfolders, right?

Point 2: Making the first and second lines (and people asked for a third line with the message preview) more distinguishable is a very tough call. I can't use a fixed colour since rows can be tagged, and if you tag with your fixed colour, then you can't read the text. TB/BB has a very complicated colour calculation that ensures good readability/contrast whatever the tag colour. Then there is a dark theme and the colour would have to be readable there, too. Bold doesn't work since it stands for unread. So I was left with italics, which personally I don't like either. The striping is done via userChrome.css, I've just uploaded a new version which you can access via the Expert Tips page (https://www.betterbird.eu/expert-tips/, https://www.betterbird.eu/expert-tips/userChrome.css). That file likely won't be compatible with a dark theme. You see, TB has two people dedicated to styling the UI, we don't have that much staff.

**Comment by Betterbird** [comment]
Oh, I've just remembered: You can style the second line any way you like via userChrome.css:
```
treechildren::-moz-tree-cell-text(colspan) {
  /* Italic is a matter of taste here. It can be removed in userChrome.css */
  font-style: normal !important;
  color: red;
  background-color: yellow;
}
```

**Comment by ynjo** [comment]
1) Yes, I follow the 18 yo bug from a long time :-D and yes, I can propagate the chosen set of columns to other folders and subfolders but something doesn't work correctly: even after I do that, I still see one line per message; to have multi-line view I need to add again the subject and to remove it, for every folder

2) The alternate colors using userChrome.css works, it's not very charming, but it does the job for now. Maybe a thin horizontal line between every message could do the job too, looking better, like Outlook does, or more padding like Postbox does.

Thanks

**Comment by Betterbird** [comment]
1. Global option will come in the next release next week.
2. For a border, add this to your userChrome.css
```
treechildren::-moz-tree-row(multiline) {
  border-bottom: 1px solid hsla(0, 0%, 50%, 0.5) !important;
}
treechildren::-moz-tree-row(multiline, colspan) {
  border-bottom: none !important;
}
```

**Comment by ynjo** [comment]
[userChrome.zip](https://github.com/Betterbird/thunderbird-patches/files/7147477/userChrome.zip)
Thanks again, using the attached css now, not bad for my taste. Maybe good to add some padding too, I'll try later

**Comment by Betterbird** [comment]
Global option and separator lines now implemented in "preview 5".

**Comment by ynjo** [comment]
Thanks! I'll try it tomorrow (at home I use Ubuntu).
In the meanwhile, I ended up using the CSS height parameter = 50px to make the multiline row a bit taller, besides coloring the subject on the second line.

**Comment by Betterbird** [comment]
There is a density setting, so 50px hard-coded won't work.

**Comment by ynjo** [comment]
But the density applies to everything, even the folder pane; instead I needed to make only the multiline row a bit taller, and with user CSS it seemed to work as expected with preview 4. I'll check the 5th.

**Comment by Betterbird** [comment]
You misunderstood: If you hard-code the height of the multiline row, that it won't follow the density setting which we coded here:
https://github.com/Betterbird/thunderbird-patches/blob/a664b921fae59ea51f4cb82a7a6b1a18299ebd44/91/features/04-feature-multiline-tree.patch#L242

**Comment by ynjo** [comment]
Ah yes, but that's what I needed, an easy way to make only the multiline part with bigger height.
Anyway the preview 5 seems to work well, thank you very much! Do you plan trying also adding the third line? You mentioned it before.

I see now that the multiline view remains in all layouts. It's not a problem for me because I use the vertical layout, but I don't know if other ppl would like it also in classical and wide view.

Last question: do you know how could I change the "unread" message properties using userChrome.css? I'd like to try make it appear like Outlook (that is a blue border on the left and a light blue background color) instead of simply bold character, to see how does it look like.

**Comment by Betterbird** [comment]
What you have now will be released as a first cut in 91.1.1, likely tomorrow, depending on TB's own progress.

The CSS for unread messages is this
```
 treechildren::-moz-tree-cell-text(unread) {
   font-weight: bold;
 }
```
The trouble is that the cell text, the individual text bits, have the "unread attribute, not the surrounding box, which is the row. So if you add a border here, you get this:
![image](https://user-images.githubusercontent.com/86832723/133393302-25146b08-0a8f-4fb2-b57d-96bc2d22835a.png)
In order to do what you want, the unread attribute would have to be repeated on the containing row. That would require code changes.

**Comment by Betterbird** [comment]
OK, use "preview 6" and this CSS:
```
treechildren::-moz-tree-row(unread) {
  border-left: 2px solid red !important;
}
treechildren::-moz-tree-row(unread, colspan) {
  border-left: none !important;
}
```
to get
![image](https://user-images.githubusercontent.com/86832723/133396479-53e500f7-8a2e-468b-ae01-785dbf0fc975.png)

**Comment by ynjo** [comment]
Good, testes preview 6 too. It solves the previous issue.
I'd like to add the left border and a background color instead of bold text for unread messages and used this CSS:

treechildren::-moz-tree-row(multiline, unread) {
  border-left: 3px solid #0080C0 !important;
  background-color: rgba(0, 128, 192, 0.1) !important;
  font-weight: normal !important;
}

treechildren::-moz-tree-row(multiline, unread, colspan) {
  border-left: none !important;
  background-color: rgba(0, 128, 192, 0.1) !important;
  font-weight: normal !important;
}

The result is the attached image
![TestPreview6](https://user-images.githubusercontent.com/43180847/133423029-7d653781-bc69-423a-b07c-ea483eeee1b2.PNG)
Looks like the bold character remains and the colspan part has a darker background color.
Dunno if it can be easily solved, anyway the result is good already, thanks a lot

**Comment by Betterbird** [comment]
Like this. As I said, the cell text is bold, not the row. And the colspan needs to undo anything the row did, since the row selector applies to it, too, so you get the effect doubled-up.
```
treechildren::-moz-tree-row(multiline, unread) {
border-left: 3px solid #0080C0 !important;
background-color: rgba(0, 128, 192, 0.1) !important;
}

treechildren::-moz-tree-row(multiline, unread, colspan) {
border-left: none !important;
background-color: none !important;
}

treechildren::-moz-tree-cell-text(multiline, unread) {
  font-weight: normal !important;
}
```

**Comment by Betterbird** [comment]
OK, there is a "preview 7" now. The change is that you don't need the compensatory CSS any more, so just:
```
treechildren::-moz-tree-row(multiline, unread) {
border-left: 3px solid #0080C0 !important;
background-color: rgba(0, 128, 192, 0.1) !important;
}

treechildren::-moz-tree-cell-text(multiline, unread) {
  font-weight: normal !important;
}
```
You should also remove this:
```
treechildren::-moz-tree-row(even, colspan) {
  background-image: linear-gradient(rgba(0,0,0,0), rgba(0,0,0,0)) !important;  /* Highlight, complication for tagged rows */
}
```
and
```
treechildren::-moz-tree-row(multiline) {
  border-bottom: 1px solid #e6e6e6 !important;
}

treechildren::-moz-tree-row(multiline, colspan) {
  border-bottom: none !important;
}
```
if you still have it. I enclose my userChrome.css again.
[userChrome.zip](https://github.com/Betterbird/thunderbird-patches/files/7170030/userChrome.zip)

**Comment by ynjo** [comment]
Yes, I removed several lines.
Here is my CSS now: [userChrome.zip](https://github.com/Betterbird/thunderbird-patches/files/7172582/userChrome.zip)
Used with preview 7. Fell free to share it, for people who like an Outlook-ish look for unread messages (maybe without the hardcoded height).

Anyway, good job, I like it!
Curious to see if there will be the possibility for a third line in the future. 
Many thanks!

**Comment by Betterbird** [comment]
Thanks. "Preview 7" has just been included into 91.1.1, so this is now official.
The third line, I'd have to see whether the message preview is available in all cases.

**Comment by ynjo** [comment]
Yes, I noticed those scenarios when the preview is available (another one is when you select a thread and you get a preview list of the messages inside that thread, if I remember correctly).
Anyway, thank you for looking into it and for the very exhaustive answer, I understand perfectly. I'm very satisfied with the work you did so far, the vertical layout makes finally sense now!

**Comment by Betterbird** [comment]
Just an early warning: This kludgy CSS
```
treechildren::-moz-tree-colspan(unread) {
  margin-left: -3px;
}
```
will need to be removed with the next release since the faulty code making it necessary was corrected:
https://github.com/Betterbird/thunderbird-patches/commit/96e6002d51b9d01089204c2f888016584abe4a08#diff-fc99e8502286ce56528458ca50c842c8aaed4fb8f39c7f64a4191a1742da9360L385

You'll notice it ;-)

**Comment by Betterbird** [comment]
This has shipped in 91.1.2-bb15 now.
